Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,1422 @@
+        self.deframer = previousDeframer
+        self.decompressor = previousState.decompressor
+      } else {
+        if let zlibMethod = Zlib.Method(encoding: decompressionAlgorithm) {
+          self.decompressor = Zlib.Decompressor(method: zlibMethod)
+        }
+        let decoder = GRPCMessageDeframer(
+          maximumPayloadSize: previousState.maximumPayloadSize,
+          decompressor: self.decompressor
+        )
+        self.deframer = NIOSingleStepByteToMessageProcessor(decoder)
+      }
+    }
+  }
+
+  struct ClientClosedServerClosedState {
+    // We still need the framer and compressor in case the server has closed
+    // but its buffer is not yet empty and still needs to send messages out to
+    // the client.
+    var framer: GRPCMessageFramer
+    var compressor: Zlib.Compressor?
+
+    // These are already deframed, so we don't need the deframer anymore.
+    var inboundMessageBuffer: OneOrManyQueue<[UInt8]>
+
+    init(previousState: ClientClosedServerOpenState) {
+      self.framer = previousState.framer
+      self.compressor = previousState.compressor
+      self.inboundMessageBuffer = previousState.inboundMessageBuffer
+      previousState.decompressor?.end()
+    }
+
+    init(previousState: ClientClosedServerIdleState) {
+      self.framer = previousState.framer
+      self.compressor = previousState.compressor
+      self.inboundMessageBuffer = previousState.inboundMessageBuffer
+      previousState.decompressor?.end()
+    }
+
+    init(previousState: ClientOpenServerClosedState) {
+      self.framer = previousState.framer
+      self.compressor = previousState.compressor
+      self.inboundMessageBuffer = previousState.inboundMessageBuffer
+      previousState.decompressor?.end()
+    }
+  }
+}
+
+// - MARK: Client
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+struct GRPCStreamStateMachine {
+  private var state: GRPCStreamStateMachineState
+  private var configuration: GRPCStreamStateMachineConfiguration
+  private var skipAssertions: Bool
+
+  init(
+    configuration: GRPCStreamStateMachineConfiguration,
+    maximumPayloadSize: Int,
+    skipAssertions: Bool = false
+  ) {
+    self.state = .clientIdleServerIdle(.init(maximumPayloadSize: maximumPayloadSize))
+    self.configuration = configuration
+    self.skipAssertions = skipAssertions
+  }
+
+  mutating func send(metadata: Metadata) throws -> HPACKHeaders {
+    switch self.configuration {
+    case .client(let clientConfiguration):
+      return try self.clientSend(metadata: metadata, configuration: clientConfiguration)
+    case .server(let serverConfiguration):
+      return try self.serverSend(metadata: metadata, configuration: serverConfiguration)
+    }
+  }
+
+  mutating func send(message: [UInt8], endStream: Bool) throws {
+    switch self.configuration {
+    case .client:
+      try self.clientSend(message: message, endStream: endStream)
+    case .server:
+      assert(
```

Comment:
Why are we not using `self.assertionFailureAndCreateRPCErrorOnInternalError` here?